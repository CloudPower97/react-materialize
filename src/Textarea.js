import React, { cloneElement, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import idgen from './idgen';
import constants from './constants';

const Textarea = ({
  id,
  children,
  className,
  s,
  m,
  l,
  xl,
  disabled,
  noLayout,
  placeholder,
  icon,
  label,
  success,
  error,
  iconClassName,
  onChange,
  validate,
  defaultValue,
  value,
  ...props
}) => {
  const _textarea = useRef(null);
  const sizes = { s, m, l, xl };

  let responsiveClasses;
  if (!noLayout) {
    responsiveClasses = { col: true };
    constants.SIZES.forEach(size => {
      responsiveClasses[size + sizes[size]] = sizes[size];
    });
  }

  const wrapperClasses = cx('input-field', responsiveClasses);

  const textareaProps = {
    placeholder,
    id,
    value,
    defaultValue,
    disabled,
    ...props
  };

  useEffect(() => {
    // eslint-disable-next-line react/prop-types
    if (props['data-length']) {
      const instance = M.CharacterCounter.init(_textarea.current);

      return () => {
        instance && instance.destroy();
      };
    }
  }, [_textarea, props]);

  useEffect(() => {
    if (_textarea.current) {
      M.textareaAutoResize(_textarea.current);
    }
  }, [value]);

  const renderLabel = () =>
    label && (
      <label
        className={cx({
          active: value || placeholder,
          'label-icon': typeof label !== 'string'
        })}
        data-success={success}
        data-error={error}
        htmlFor={id}
      >
        {label}
      </label>
    );

  const renderHelper = () =>
    [error || success] && (
      <span className="helper-text" data-error={error} data-success={success} />
    );

  const renderIcon = () => {
    if (!icon) return;

    return cloneElement(icon, { className: 'prefix' });
  };

  return (
    <div className={wrapperClasses}>
      {renderIcon()}
      <textarea
        {...textareaProps}
        ref={_textarea}
        onChange={onChange}
        className={cx('materialize-textarea', { validate }, className)}
      />
      {renderLabel()}
      {renderHelper()}
      {children}
    </div>
  );
};

Textarea.propTypes = {
  children: PropTypes.node,
  /*
   * Strip away all layout classes such as col and sX
   */
  noLayout: PropTypes.bool,
  /*
   * Responsive size for Mobile Devices
   */
  s: PropTypes.number,
  /*
   * Responsive size for Tablet Devices
   */
  m: PropTypes.number,
  /*
   * Responsive size for Desktop Devices
   */
  l: PropTypes.number,
  /**
   *  Responsive size for Large Desktop Devices
   */
  xl: PropTypes.number,
  /*
   * disabled input
   */
  disabled: PropTypes.bool,
  /*
   * Placeholder string
   */
  placeholder: PropTypes.string,
  /*
   * override id
   * @default idgen()
   */
  id: PropTypes.string,
  /*
   * render icon next to input
   */
  icon: PropTypes.node,
  /*
   * icon classname
   */
  iconClassName: PropTypes.string,
  /*
   * textarea label
   */
  label: PropTypes.string,
  /*
   * Input initial value
   */
  defaultValue: PropTypes.string,
  /*
   * predefined value
   */
  value: PropTypes.string,
  /*
   * Add validate class to input
   */
  validate: PropTypes.bool,
  /*
   * Custom success message
   */
  success: PropTypes.string,
  /*
   * Custom error message
   */
  error: PropTypes.string,
  /*
   * textarea classname
   */
  className: PropTypes.string,
  /*
   * onChange callback
   */
  onChange: PropTypes.func
};

Textarea.defaultProps = {
  id: `textarea_${idgen()}`
};

export default Textarea;
