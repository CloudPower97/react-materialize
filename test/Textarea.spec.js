import React from 'react';
import { shallow, mount } from 'enzyme';
import Textarea from '../src/Textarea';

describe('<Textarea />', () => {
  let wrapper;

  test('Textarea', () => {
    wrapper = shallow(<Textarea />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Handles layout classes', () => {
    wrapper = shallow(<Textarea s={4} m={6} />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeTruthy();
    expect(wrapper.hasClass('s4')).toBeTruthy();
    expect(wrapper.hasClass('m6')).toBeTruthy();
  });

  test('Handles removing layout classes', () => {
    wrapper = shallow(<Textarea s={4} m={6} noLayout />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeFalsy();
    expect(wrapper.hasClass('s4')).toBeFalsy();
    expect(wrapper.hasClass('m6')).toBeFalsy();
  });

  test('Disabled', () => {
    wrapper = shallow(<Textarea value="I was here" disabled />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Placeholder', () => {
    wrapper = shallow(<Textarea placeholder="I was here" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Label', () => {
    wrapper = shallow(<Textarea label="Write something here..." />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Icon', () => {
    wrapper = shallow(<Textarea icon="mode_edit" iconClassName="red" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with custom className', () => {
    wrapper = shallow(<Textarea className="custom-class" />);
    const textareaElement = wrapper.find('textarea');
    expect(textareaElement.hasClass('materialize-textarea')).toBeTruthy();
    expect(textareaElement.hasClass('custom-class')).toBeTruthy();
  });

  describe('undefined M', () => {
    let __M;
    beforeEach(() => {
      __M = global.M;
      global.M = undefined;
    });
    afterEach(() => {
      global.M = __M;
    });
    test('doesnt throw without M', () => {
      const element = mount(<Textarea />);
      expect(() => element.setProps({ label: 'foo' })).not.toThrow();
    });
  });
});
