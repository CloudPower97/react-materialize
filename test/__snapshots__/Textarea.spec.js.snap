// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea /> Disabled 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    disabled={true}
    id="textarea_mockId"
    value="I was here"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Icon 1`] = `
<div
  className="input-field col"
>
  <Component
    className="prefix"
  />
  <textarea
    className="materialize-textarea"
    id="textarea_mockId"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Label 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="textarea_mockId"
  />
  <label
    className=""
    htmlFor="textarea_mockId"
  >
    Write something here...
  </label>
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Placeholder 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="textarea_mockId"
    placeholder="I was here"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Textarea 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="textarea_mockId"
  />
  <span
    className="helper-text"
  />
</div>
`;
